parameterize Block by the header type

JaredCorduan opened the pull request:
This changes the `Block` type to be parameterized by the block header type. I did not attempt to make the tests more generic, but just used the TPraos `BHeader` directly.

closes #2459 

EDIT -  I've also now:

* Moved `TxIn` and `TxId` to cardano-ledger-core
* Moved `BlocksMade` to cardano-ledger-core (I placed it in `BaseTypes`)
* Moved `Block` to cardano-ledger-core
* Moved `LaxBlock` to the shelley tests
* The signal for `BBODY` is once again a `Block`, but now we use the `BHeaderView` for the block header in the `BBODY` rule. There is one awkward snag with this (but I think it is worth it), the `Block` pattern requires a `ToCBOR` instance on the block header. I didn't want to add a `ToCBOR` for the BHeaderView, since we don't actually ever want it to be serialized, so when giving `BBODY` a `Block`, we need to use the constructor `Block'` with something bogus in the bytes field.

Commit:
parametrize Block by the header type

nc6 commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/BlockChain.hs:
```diff
@@ -250,27 +250,33 @@ bbHash (TxSeq' _ bodies wits md) =
 
-data Block era
-  = Block' !(TP.BHeader (Crypto era)) !(Era.TxSeq era) BSL.ByteString
+data Block h era
```
Now that `Block` is parametric over the header type, I wonder whether we should move it to `cardano-ledger-core`

JaredCorduan replied:
These are some great ideas, thanks @nc6 ! I'll do some experimenting. :scientist:

JaredCorduan replied:
@nc6 this turned out mostly great, see my notes in the PR notes at the top.

nc6 reviewed (changes requested):
This looks great, but now that `Block` is parametrised by the header, can we revert to using `Block` as the signal for `BBODY` rather than the tuple?

Commit:
BBODY rule to use a Block as a signal again

Commit:
move TxIn and TxId to core

Commit:
move BlocksMade to core (BaseTypes)

Commit:
move Block to core

Additionally, move LaxBlock to testing

nc6 commented on libs/cardano-ledger-core/src/Cardano/Ledger/Block.hs:
```diff
@@ -0,0 +1,154 @@
+import NoThunks.Class (NoThunks (..))
+
+data Block h era
+  = Block' !(h (Crypto era)) !(Era.TxSeq era) BSL.ByteString
```
Rather than exposing the constructor directly, how about having a pair of patterns which we use in cases where we don't care about serialisation: a unidirectional pattern `UnserialisedBlock` that allows us to pull out the header and the TxSeq, and a bidirectional `UnsafeUnserialisedBlock` which constructs the block with an error in the bytestring? This way it's a bit clearer what's going on, and we don't have to expose the potentially unsafe constructor?

JaredCorduan replied:
Absolutely, that's much better!

nc6 reviewed (changes requested):
Looks great, I think the changes make this much nicer! I have left one final suggestion. Also, is this file intentionally checked in?
- eras/shelley/impl/dist/cabal-config-flags

nc6 committed:
Introduce 'UnserialisedBlock' patterns.

These make it clear when we are accessing the block without its
serialisation. Unfortunately we cannot quite hide the Block'
constructor, since it's used by `LaxBlock`. Maybe we should rename it to
`UnsafeBlock`, though.

We also add 'UnsafeUnserialisedBlock', which allows construction of a
block without the ability to serialise it.

JaredCorduan commented:
> Looks great, I think the changes make this much nicer! I have left one final suggestion. Also, is this file intentionally checked in?
> 
> * eras/shelley/impl/dist/cabal-config-flags

oops! I'll remove it, thanks! Edit: done.

Commit:
Merge branch 'master' into jc/parametrize-block-by-header
